File: packages/python-sdk/src/treedb_sdk/adapters/federation_internal.py
from __future__ import annotations

from typing import Any

from .common import json_request, segment
from treedb_sdk.transport import Transport


class FederationInternalAdapter:
    def __init__(self, transport: Transport) -> None:
        self.transport = transport

    def health(self) -> Any: return json_request(self.transport, "GET", "/api/v1/internal/federation/health")
    def proxy(self, body: Any) -> Any: return json_request(self.transport, "POST", "/api/v1/internal/federation/proxy", body)
    def export_mirror(self, repo_id: str, body: Any) -> Any: return json_request(self.transport, "POST", f"/api/v1/internal/federation/repos/{segment(repo_id)}/mirror/export", body)
    def import_mirror(self, repo_id: str, body: Any) -> Any: return json_request(self.transport, "POST", f"/api/v1/internal/federation/repos/{segment(repo_id)}/mirror/import", body)

File: packages/rust-sdk/src/adapters/federation_internal.rs
use std::sync::Arc;

use serde_json::Value;

use crate::adapters::common::{json_request, segment};
use crate::error::TreeDxResult;
use crate::transport::{Transport, TreeDxHttpMethod};

#[derive(Clone)]
pub struct FederationInternalAdapter {
    transport: Arc<dyn Transport>,
}

impl FederationInternalAdapter {
    pub fn new(transport: Arc<dyn Transport>) -> Self {
        Self { transport }
    }

    pub async fn health(&self) -> TreeDxResult<Value> {
        json_request(
            &self.transport,
            TreeDxHttpMethod::Get,
            "/api/v1/internal/federation/health",
            None,
            None,
        )
        .await
    }
    pub async fn proxy(&self, body: Value) -> TreeDxResult<Value> {
        json_request(
            &self.transport,
            TreeDxHttpMethod::Post,
            "/api/v1/internal/federation/proxy",
            Some(body),
            None,
        )
        .await
    }
    pub async fn export_mirror(&self, repo_id: &str, body: Value) -> TreeDxResult<Value> {
        json_request(
            &self.transport,
            TreeDxHttpMethod::Post,
            format!(
                "/api/v1/internal/federation/repos/{}/mirror/export",
                segment(repo_id)
            ),
            Some(body),
            None,
        )
        .await
    }
    pub async fn import_mirror(&self, repo_id: &str, body: Value) -> TreeDxResult<Value> {
        json_request(
            &self.transport,
            TreeDxHttpMethod::Post,
            format!(
                "/api/v1/internal/federation/repos/{}/mirror/import",
                segment(repo_id)
            ),
            Some(body),
            None,
        )
        .await
    }
}
